File: tools/create_test_dmd.py
#!/usr/bin/env python3
"""
Create a test 128x32 DMD animation for validating CrabCrust conversion pipeline
Classic pinball DMD style with "JACKPOT" animation
"""

from PIL import Image, ImageDraw, ImageFont
import os

# DMD dimensions
WIDTH = 128
HEIGHT = 32
FRAMES = 12

def create_frame(frame_num):
    """Create a single frame of the animation"""
    img = Image.new('RGB', (WIDTH, HEIGHT), color=(0, 0, 0))
    draw = ImageDraw.Draw(img)

    # Classic DMD orange color
    color = (255, 128, 0)

    # Animation phases
    if frame_num < 3:
        # Flash full screen
        if frame_num % 2 == 0:
            draw.rectangle([(0, 0), (WIDTH, HEIGHT)], fill=color)
    elif frame_num < 6:
        # Draw border
        thickness = 2
        draw.rectangle([(0, 0), (WIDTH-1, HEIGHT-1)], outline=color, width=thickness)
        # Add "JACKPOT" text
        text = "JACKPOT"
        # Calculate text position (centered)
        bbox = draw.textbbox((0, 0), text)
        text_width = bbox[2] - bbox[0]
        text_height = bbox[3] - bbox[1]
        x = (WIDTH - text_width) // 2
        y = (HEIGHT - text_height) // 2 - 2
        draw.text((x, y), text, fill=color)
    else:
        # Animate stars/sparkles around text
        import random
        random.seed(frame_num)

        # Draw border
        draw.rectangle([(0, 0), (WIDTH-1, HEIGHT-1)], outline=color, width=2)

        # Draw "JACKPOT" text
        text = "JACKPOT"
        bbox = draw.textbbox((0, 0), text)
        text_width = bbox[2] - bbox[0]
        text_height = bbox[3] - bbox[1]
        x = (WIDTH - text_width) // 2
        y = (HEIGHT - text_height) // 2 - 2
        draw.text((x, y), text, fill=color)

        # Add random sparkles
        for _ in range(10):
            sx = random.randint(2, WIDTH-3)
            sy = random.randint(2, HEIGHT-3)
            draw.point((sx, sy), fill=color)
            # Make some sparkles bigger
            if random.random() > 0.5:
                draw.point((sx+1, sy), fill=color)
                draw.point((sx, sy+1), fill=color)

    return img

def main():
    print("🎰 Creating test DMD animation (128x32)...")

    # Create frames
    frames = []
    for i in range(FRAMES):
        print(f"   Frame {i+1}/{FRAMES}")
        frame = create_frame(i)
        frames.append(frame)

    # Save as animated GIF
    output_path = "/home/user/CrabCrust/ref/test_jackpot_dmd.gif"
    frames[0].save(
        output_path,
        save_all=True,
        append_images=frames[1:],
        duration=100,  # 100ms per frame
        loop=0  # Loop forever
    )

    print(f"\n✨ Created: {output_path}")
    print(f"   Dimensions: {WIDTH}x{HEIGHT}")
    print(f"   Frames: {FRAMES}")
    print(f"   Duration: {FRAMES * 100}ms")
    print(f"\n💡 Test conversion:")
    print(f"   crabcrust convert {output_path} --play --save jackpot")

if __name__ == "__main__":
    main()

File: tools/create_test_dmd.rs
// Create a test 128x32 DMD animation for validating CrabCrust conversion pipeline
// Classic pinball DMD style with "JACKPOT" animation

use std::fs::File;
use std::io::BufWriter;

const WIDTH: u32 = 128;
const HEIGHT: u32 = 32;
const FRAMES: usize = 12;

fn create_frame(frame_num: usize) -> Vec<u8> {
    let mut pixels = vec![0u8; (WIDTH * HEIGHT * 3) as usize];

    // Classic DMD orange color
    let color = [255u8, 128u8, 0u8];

    // Animation phases
    if frame_num < 3 {
        // Flash full screen on even frames
        if frame_num % 2 == 0 {
            for chunk in pixels.chunks_mut(3) {
                chunk.copy_from_slice(&color);
            }
        }
    } else if frame_num < 6 {
        // Draw border and "JACKPOT" text (simplified - just border for now)
        // Top and bottom borders
        for x in 0..WIDTH {
            for y in 0..2 {
                let idx = ((y * WIDTH + x) * 3) as usize;
                pixels[idx..idx + 3].copy_from_slice(&color);
            }
            for y in (HEIGHT - 2)..HEIGHT {
                let idx = ((y * WIDTH + x) * 3) as usize;
                pixels[idx..idx + 3].copy_from_slice(&color);
            }
        }
        // Left and right borders
        for y in 0..HEIGHT {
            for x in 0..2 {
                let idx = ((y * WIDTH + x) * 3) as usize;
                pixels[idx..idx + 3].copy_from_slice(&color);
            }
            for x in (WIDTH - 2)..WIDTH {
                let idx = ((y * WIDTH + x) * 3) as usize;
                pixels[idx..idx + 3].copy_from_slice(&color);
            }
        }

        // Draw simple "JACKPOT" pattern (simplified vertical bars)
        let bars = [20, 35, 50, 65, 80, 95, 110];
        for &bar_x in &bars {
            for y in 12..20 {
                for x in bar_x..(bar_x + 3).min(WIDTH) {
                    let idx = ((y * WIDTH + x) * 3) as usize;
                    pixels[idx..idx + 3].copy_from_slice(&color);
                }
            }
        }
    } else {
        // Sparkle animation - border plus random dots
        // Draw border
        for x in 0..WIDTH {
            for y in 0..2 {
                let idx = ((y * WIDTH + x) * 3) as usize;
                pixels[idx..idx + 3].copy_from_slice(&color);
            }
            for y in (HEIGHT - 2)..HEIGHT {
                let idx = ((y * WIDTH + x) * 3) as usize;
                pixels[idx..idx + 3].copy_from_slice(&color);
            }
        }
        for y in 0..HEIGHT {
            for x in 0..2 {
                let idx = ((y * WIDTH + x) * 3) as usize;
                pixels[idx..idx + 3].copy_from_slice(&color);
            }
            for x in (WIDTH - 2)..WIDTH {
                let idx = ((y * WIDTH + x) * 3) as usize;
                pixels[idx..idx + 3].copy_from_slice(&color);
            }
        }

        // Text bars
        let bars = [20, 35, 50, 65, 80, 95, 110];
        for &bar_x in &bars {
            for y in 12..20 {
                for x in bar_x..(bar_x + 3).min(WIDTH) {
                    let idx = ((y * WIDTH + x) * 3) as usize;
                    pixels[idx..idx + 3].copy_from_slice(&color);
                }
            }
        }

        // Add sparkles (pseudo-random based on frame number)
        let seed = frame_num * 123456789;
        for i in 0..15 {
            let x = ((seed + i * 9876) % WIDTH as usize) as u32;
            let y = ((seed + i * 6789) % HEIGHT as usize) as u32;
            if x >= 4 && x < WIDTH - 4 && y >= 4 && y < HEIGHT - 4 {
                let idx = ((y * WIDTH + x) * 3) as usize;
                pixels[idx..idx + 3].copy_from_slice(&color);
            }
        }
    }

    pixels
}

fn main() -> Result<(), Box<dyn std::error::Error>> {
    println!("🎰 Creating test DMD animation (128x32)...");

    let output_path = "/home/user/CrabCrust/ref/test_jackpot_dmd.gif";

    // Create frames
    let mut frames = Vec::new();
    for i in 0..FRAMES {
        println!("   Frame {}/{}", i + 1, FRAMES);
        let pixels = create_frame(i);
        frames.push(pixels);
    }

    // For now, just save the first frame as a simple PPM
    // (GIF encoding requires external library - let's use a simpler format first)
    let ppm_path = "/home/user/CrabCrust/ref/test_jackpot_dmd.ppm";
    let file = File::create(ppm_path)?;
    let mut writer = BufWriter::new(file);

    use std::io::Write;
    write!(writer, "P6\n{} {}\n255\n", WIDTH, HEIGHT)?;
    writer.write_all(&frames[5])?; // Use middle frame

    println!("\n✨ Created: {}", ppm_path);
    println!("   Dimensions: {}x{}", WIDTH, HEIGHT);
    println!("   Format: PPM (will convert to GIF)");

    Ok(())
}
